tests/test_backend.py

```diff
@@ -462,26 +454,62 @@ def test_glob(tmpdir, files, pattern, folder, expected, backend):
         ),
     ]
 )
-def test_ls(tmpdir, path, content, expected_content, backend):
+def test_ls(tmpdir, backend):
 
-    backend_path = backend.join(
+    prefix = backend.join(
         pytest.ID,
         'test_ls',
-        path,
     )
 
-    for file in content:
-        local_file = os.path.join(tmpdir, file)
-        audeer.mkdir(os.path.dirname(local_file))
-        audeer.touch(local_file)
-        remote_file = backend.join(backend_path, file)
+    sub_content = [  # two versions of same file
+        (f'{prefix}/sub/file.txt', None, '1.0.0'),
+        (f'{prefix}/sub/file.txt', None, '2.0.0'),
+    ]
+    sub_content_latest = sub_content[-1:]
+
+    content = [  # three files with different extensions
+        (f'{prefix}/file.tar.gz', '', '1.0.0'),
+        (f'{prefix}/file.tar.gz', None, '1.0.0'),
+        (f'{prefix}/file.tar.gz', '.tar.gz', '1.0.0'),
+    ]
+    content_latest = content
+
+    content = sub_content + content
+    content_latest = content_latest + sub_content_latest
+
+    # create content
+
+    tmp_file = os.path.join(tmpdir, '~')
+    for path, ext, version in content:
+        audeer.touch(tmp_file)
         backend.put_file(
-            local_file,
-            remote_file,
-            '1.0.0',
+            tmp_file,
+            path,
+            version,
+            ext=ext,
         )
 
-    assert backend.ls(backend_path) == expected_content
+    # test
+
+    for folder, expected, expected_latest in [
+        ('/', content, content_latest),
+        ('sub', sub_content, sub_content_latest),
+    ]:
+        folder = backend.join(prefix, folder)
+
+        expected = [  # replace ext where it is None
```

Comment:
Maybe

```suggestion
        expected = [  # remove ext where it is None
```

Reply:
If the file extension is `None` we replace it with the default file extension (i.e. characters after last dot), so I think the term "replace" is correct here.

Reply:
You are right.
I just tought `p.split('.')[-1]` looks not like replacement, but if you read carefully you understand it.